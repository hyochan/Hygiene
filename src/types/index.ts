import { StyleProp, TextStyle } from 'react-native';

import { SFC } from 'react';
import firebase from 'firebase/app';

export type FirebaseUser = firebase.User;

export enum SocialAuthProvider {
  Google = 'google',
  Facebook = 'facebook',
  Apple = 'apple',
};

export type User = {
  uid: string;
  displayName: string | null;
  email: string;
  introduction?: string | null;
  point: number;
  photoURL: string | null;
  thumbURL?: string;
  signInMethod?: string | null;
}

export type Reply = {
  id: string;
  text: string;
  writerId: string;
  createdAt?: Date;
  updatedAt?: Date;
}

interface IconProps {
  style?: StyleProp<TextStyle>;
  width?: number | string;
  height?: number | string;
  children?: never;
}

export type IconType = SFC<IconProps>;

export interface Activity {
  id: string;
  writerId: string;
  message: string;
  urls: string[];
  category: string;
  createdAt?: Date;
  updatedAt?: Date;
}

export interface Photo {
  uri: string;
  thumbUri?: string;
  added?: boolean;
  deleted?: boolean;
}

export enum CategoryType {
  HandWash = 'HandWash',
  WearMask = 'WearMask',
  StayHome = 'StayHome',
  GoodConsumption = 'GoodConsumption',
  Etc = 'Etc',
}

export interface Category {
  type: string;
  point: number;
  text: string;
  svgIcon: any;
  color: string;
}

export enum SelectImageActionType {
  LAUNCH_CAMERA = 0,
  LAUNCH_GALLERY = 1,
  CANCEL = 2,
}

export enum SelectMyFeedMoreActionType {
  DELETE = 0,
  CANCEL = 1,
}

export enum SelectPeerFeedMoreActionType {
  CANCEL = 0,
}

export interface Throphy {
  type: string;
  text: string;
  count: number;
}
