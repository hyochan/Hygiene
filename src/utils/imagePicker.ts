import * as ImageManipulator from 'expo-image-manipulator';
import * as ImagePicker from 'expo-image-picker';
import * as Permissions from 'expo-permissions';

import { Image, Platform } from 'react-native';

export const THUMBNAIL_SIZES = {
  MAX_WIDTH: 192,
  MAX_HEIGHT: 192,
};

export interface ImageSize {
  width: number;
  height: number;
}

export const getOriginalImageSize = async (imageUri: string): Promise<ImageSize> => new Promise<ImageSize>(
  (resolve, reject) => Image.getSize(
    imageUri,
    (width: number, height: number) => resolve({
      width,
      height,
    }),
    reject,
  ),
);

export const resizeImage = async ({
  imageUri,
  maxWidth,
  maxHeight,
}: {
  imageUri: string;
  maxWidth?: number;
  maxHeight?: number;
}): Promise<ImageManipulator.ImageResult> => {
  const manipulate = ({
    width,
    height,
  }: {
    width?: number;
    height?: number;
  }): Promise<ImageManipulator.ImageResult> => ImageManipulator.manipulateAsync(
    imageUri,
    [{ resize: { width, height } }],
    { compress: 1, format: ImageManipulator.SaveFormat.PNG },
  );

  const { width: originWidth, height: originHeight } = await getOriginalImageSize(imageUri);

  if (originWidth >= originHeight) {
    if (maxWidth === undefined || maxWidth === null) {
      if (originWidth < THUMBNAIL_SIZES.MAX_WIDTH) {
        return manipulate({ width: originWidth });
      }

      return manipulate({ width: THUMBNAIL_SIZES.MAX_WIDTH });
    }

    if (originWidth < maxWidth) {
      return manipulate({ width: originWidth });
    }

    return manipulate({ width: maxWidth });
  }

  if (maxHeight === undefined || maxHeight === null) {
    if (originHeight < THUMBNAIL_SIZES.MAX_HEIGHT) {
      return manipulate({ height: originHeight });
    }

    return manipulate({ height: THUMBNAIL_SIZES.MAX_HEIGHT });
  }

  if (originHeight < maxHeight) {
    return manipulate({ height: originHeight });
  }

  return manipulate({ height: maxHeight });
};

enum MediaTypeOptions {
  All = 'All',
  Videos = 'Videos',
  Images = 'Images',
}

enum PermissionStatus {
  UNDETERMINED = 'undetermined',
  GRANTED = 'granted',
  DENIED = 'denied',
}

const photoOptions = {
  mediaTypes: ImagePicker.MediaTypeOptions.Images,
  allowsEditing: true,
  exif: true,
};

const requestPermissions = async (
  type: string,
): Promise<Permissions.PermissionStatus> => {
  if (type === 'photo') {
    const { status } = await Permissions.askAsync(Permissions.CAMERA_ROLL);

    return status;
  }

  const { status } = await Permissions.askAsync(Permissions.CAMERA);

  return status;
};

export const launchCameraAsync = async (): Promise<ImagePicker.ImagePickerResult | null> => {
  const permissionStatus = await requestPermissions('camera');

  if (permissionStatus === PermissionStatus.GRANTED) {
    return ImagePicker.launchCameraAsync(photoOptions);
  }

  return null;
};

export const launchImageLibraryAsync = async (): Promise<ImagePicker.ImagePickerResult | null> => {
  if (Platform.OS === 'web') {
    return ImagePicker.launchImageLibraryAsync(photoOptions);
  }

  const permissionStatus = await requestPermissions('photo');

  if (permissionStatus === PermissionStatus.GRANTED) {
    return ImagePicker.launchImageLibraryAsync(photoOptions);
  }

  return null;
};
