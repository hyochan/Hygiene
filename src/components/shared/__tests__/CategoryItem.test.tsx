import 'react-native';

import * as React from 'react';

import {
  RenderAPI,
  render,
} from '@testing-library/react-native';
import { createTestElement, createTestProps } from '../../../../test/testUtils';

import { CategoryType } from '../../../types';
import Shared from '../CategoryItem';
import { categoryMap } from '../../../utils/constants';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
let props: any;
let component: React.ReactElement;
let testingLib: RenderAPI;

describe('Rendering', () => {
  beforeEach(() => {
    props = createTestProps({
      category: categoryMap[CategoryType.HandWash],
    });

    component = createTestElement(<Shared {...props} />);
    testingLib = render(component);
  });

  it('renders without crashing', () => {
    const baseElement = testingLib.toJSON();

    expect(baseElement).toMatchSnapshot();
    expect(baseElement).toBeTruthy();
  });
});

describe('Interaction', () => {
  beforeEach(() => {
    testingLib = render(component);
  });

  it('should simulate onClick', () => {
    const baseElement = testingLib.toJSON();

    expect(baseElement).toMatchSnapshot();
    // const btn = testingLib.queryByTestId('btn');
    // act(() => {
    //   fireEvent.press(btn);
    // });
    // expect(cnt).toBe(3);
  });
});
