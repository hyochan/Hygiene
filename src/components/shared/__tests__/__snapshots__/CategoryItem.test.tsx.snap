// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interaction should simulate onClick 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "column",
        "height": 136,
        "justifyContent": "center",
        "marginLeft": 20,
        "opacity": 1,
        "width": 96,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderColor": "#E6E6E6",
            "borderRadius": 16,
            "borderStyle": "solid",
            "borderWidth": 1,
            "height": 120,
            "justifyContent": "center",
            "shadowColor": "rgba(0, 0, 0, 0.06)",
            "shadowOffset": Object {
              "height": 2,
              "width": 2,
            },
            "shadowOpacity": 1,
            "shadowRadius": 4,
            "width": 96,
          },
        ]
      }
    >
      <SvgMock
        fill="black"
        height={48}
        width={48}
      />
      <Text
        style={
          Object {
            "color": "#02C7D3",
            "fontSize": 18,
            "position": "absolute",
            "right": 12,
            "top": 12,
          }
        }
      >
        +
        3
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "borderRadius": 16,
            "bottom": 0,
            "minHeight": 30,
            "minWidth": 57,
            "paddingBottom": 0,
            "paddingLeft": 8,
            "paddingRight": 8,
            "paddingTop": 0,
            "position": "absolute",
          },
          Object {
            "backgroundColor": "#02C7D3",
            "justifyContent": "center",
          },
        ]
      }
    >
      <Text
        numberOfLines={2}
        style={
          Object {
            "color": "white",
            "textAlign": "center",
          }
        }
      >
        Washing hands
      </Text>
    </View>
  </View>
</View>
`;

exports[`Rendering renders without crashing 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "column",
        "height": 136,
        "justifyContent": "center",
        "marginLeft": 20,
        "opacity": 1,
        "width": 96,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderColor": "#E6E6E6",
            "borderRadius": 16,
            "borderStyle": "solid",
            "borderWidth": 1,
            "height": 120,
            "justifyContent": "center",
            "shadowColor": "rgba(0, 0, 0, 0.06)",
            "shadowOffset": Object {
              "height": 2,
              "width": 2,
            },
            "shadowOpacity": 1,
            "shadowRadius": 4,
            "width": 96,
          },
        ]
      }
    >
      <SvgMock
        fill="black"
        height={48}
        width={48}
      />
      <Text
        style={
          Object {
            "color": "#02C7D3",
            "fontSize": 18,
            "position": "absolute",
            "right": 12,
            "top": 12,
          }
        }
      >
        +
        3
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "borderRadius": 16,
            "bottom": 0,
            "minHeight": 30,
            "minWidth": 57,
            "paddingBottom": 0,
            "paddingLeft": 8,
            "paddingRight": 8,
            "paddingTop": 0,
            "position": "absolute",
          },
          Object {
            "backgroundColor": "#02C7D3",
            "justifyContent": "center",
          },
        ]
      }
    >
      <Text
        numberOfLines={2}
        style={
          Object {
            "color": "white",
            "textAlign": "center",
          }
        }
      >
        Washing hands
      </Text>
    </View>
  </View>
</View>
`;
